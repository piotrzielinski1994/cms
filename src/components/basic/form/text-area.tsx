import { cn } from '@/utils/tailwind';
import { TextareaHTMLAttributes } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import Form from './root/form';
import { classNames } from './text-input';

type TextAreaProps = Omit<TextareaHTMLAttributes<HTMLTextAreaElement>, 'name'> & {
  name: string;
  error?: string;
};

type TextAreaContainerProps<T extends FieldValues> = Omit<TextAreaProps, 'name'> & {
  control: Control<T>;
  name: Path<T>;
};

const TextArea = ({ error, className, ...props }: TextAreaProps) => {
  return (
    <>
      <textarea
        {...props}
        value={props.value ?? ''}
        className={cn(
          ...classNames({ isValid: !error }),
          'min-h-[calc(2*1px+2*0.5rem+3*1.5rem)]',
          className,
        )}
      />
      <Form.Error>{error}</Form.Error>
    </>
  );
};

const TextAreaContainer = <T extends FieldValues>(props: TextAreaContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <TextArea error={fieldState.error?.message} {...rest} {...field} />;
};

export { TextArea, TextAreaContainer };
