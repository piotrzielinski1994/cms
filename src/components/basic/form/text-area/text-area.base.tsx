import { forwardRef, InputHTMLAttributes } from 'react';
import Form from '../root/form';

type InputProps = Omit<InputHTMLAttributes<HTMLTextAreaElement>, 'value'> & {
  name: string;
  value?: string;
};

const Input = forwardRef<HTMLTextAreaElement, InputProps>((props, ref) => {
  return <textarea ref={ref} contentEditable={false} {...props} value={props.value ?? ''} />;
});

const TextAreaBase = {
  Root: Form.Group,
  Label: Form.Label,
  Input,
  Error: Form.Error,
};

export default TextAreaBase;
