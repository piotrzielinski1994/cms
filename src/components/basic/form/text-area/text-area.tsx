import { EnhancedHtmlProps } from '@/utils/html/html.types';
import { cn } from '@/utils/tailwind';
import { BoolMap } from '@/utils/types';
import { forwardRef, ReactNode } from 'react';
import Form from '../root/form';
import { styles as textInputStyles } from '../text-input/text-input';

type TextAreaProps = NativeProps & {
  label?: ReactNode;
  error?: string;
};

// prettier-ignore
type NativeProps = EnhancedHtmlProps<'textarea', {
  name: string;
  value?: string;
}>;

const styles = {
  native: ({ isValid, isDisabled }: BoolMap<'isValid' | 'isDisabled'>) =>
    cn(textInputStyles.native({ isValid }), 'min-h-[calc(2*1px+2*0.5rem+3*1.5rem)]', {
      'resize-none': isDisabled,
    }),
};

const Component = forwardRef<HTMLTextAreaElement, TextAreaProps>((props, ref) => {
  const { label, error, className, ...rest } = props;
  return (
    <Root className={className}>
      {label && <Label htmlFor={rest.id}>{label}</Label>}
      <Input ref={ref} aria-invalid={!!error} {...rest} />
      <Error>{error}</Error>
    </Root>
  );
});

const Input = forwardRef<HTMLTextAreaElement, NativeProps>((props, ref) => {
  const { value = '', className, ...rest } = props;
  const base = styles.native({
    isValid: !rest['aria-invalid'],
    isDisabled: !!props.disabled,
  });

  return (
    <textarea
      ref={ref}
      contentEditable={false}
      {...rest}
      className={cn(base, className)}
      value={value}
    />
  );
});

const Root = Form.Group;
const Label = Form.Label;
const Error = Form.Error;

const TextArea = Object.assign(Component, { Root, Label, Input, Error });

export { TextArea };
