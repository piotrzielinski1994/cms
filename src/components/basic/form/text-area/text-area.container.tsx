import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { TextArea } from './text-area';

type TextAreaContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<ComponentProps<typeof TextArea>, 'name'>;

const TextAreaContainer = <T extends FieldValues>(props: TextAreaContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <TextArea error={fieldState.error?.message} {...rest} {...field} />;
};

export { TextAreaContainer };
