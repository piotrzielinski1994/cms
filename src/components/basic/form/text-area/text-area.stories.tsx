import Form from '@/components/basic/form/root/form';
import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useState, type ComponentProps } from 'react';
import { TextArea as TextAreaComponent } from './text-area';

type Args = ComponentProps<typeof TextAreaComponent> & {
  label?: string;
};

const ControlledInput = (props: Args) => {
  const [value, setValue] = useState('');
  return (
    <TextAreaComponent {...props} onChange={(e) => setValue(e.currentTarget.value)} value={value} />
  );
};

const meta: Meta<ComponentProps<typeof ControlledInput>> = {
  component: ControlledInput,
  title: 'Components/Basic/Form/TextArea',
  argTypes: {
    label: { control: 'text' },
    placeholder: { control: 'text' },
    error: { control: 'text' },
    disabled: { control: 'boolean' },
  },
  args: {
    label: DEFAULT_VALUE,
    placeholder: DEFAULT_VALUE,
    disabled: false,
  },
};

const Render = ({ label, placeholder, ...args }: Args) => {
  const t = useTranslations('storybook.basic.form');
  const t2 = useTranslations('storybook.basic.form.textInput');

  return (
    <div className="grid gap-2">
      <Form.Group>
        <Form.Label htmlFor="input1">{getFallback(label, t('textArea'))}</Form.Label>
        <ControlledInput
          {...args}
          placeholder={getFallback(placeholder, t2('default.placeholder'))}
          id="input1"
          name="input1"
        />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input2">{t2('disabled.label')}</Form.Label>
        <TextAreaComponent
          id="input2"
          name="input2"
          placeholder={t2('disabled.placeholder')}
          disabled
        />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input3">{t2('invalid.label')}</Form.Label>
        <TextAreaComponent
          {...args}
          id="input3"
          name="input3"
          value={t2('invalid.value')}
          error={t2('invalid.error')}
        />
      </Form.Group>
    </div>
  );
};

const TextArea: StoryObj<typeof TextAreaComponent> = { render: Render };

export { TextArea };
export default meta;
