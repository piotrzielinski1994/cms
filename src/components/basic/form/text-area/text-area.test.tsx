import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { TextArea } from './text-area';

describe('TextArea', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    label: 'Label',
    value: 'value',
    onChange: vi.fn(),
  } satisfies ComponentProps<typeof TextArea>;

  it('should have no accessibility violations', async () => {
    const { container } = render(<TextArea {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<TextArea {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
