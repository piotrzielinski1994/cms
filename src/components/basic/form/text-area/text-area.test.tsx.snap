// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea > should match the snapshot 1`] = `
<div>
  <div
    class="grid gap-1 content-start grid-rows-[auto_1fr]"
  >
    <label
      class="self-end"
      for="id"
    >
      Label
    </label>
    <textarea
      class="p-2 border border-solid border-current bg-input placeholder-foreground/50 [&:enabled]:hover:border-foreground/90 [&:enabled]:hover:ring-foreground/90 disabled:cursor-not-allowed disabled:text-foreground/50 outline-none ring-inset focus-within:ring-1 ring-current min-h-[calc(2*1px+2*0.5rem+3*1.5rem)]"
      contenteditable="false"
      id="id"
      name="name"
    >
      value
    </textarea>
    <span
      class="min-h-[1em] text-sm text-red-500 leading-none"
      role="alert"
    />
  </div>
</div>
`;
